pkg/crd/gen.go

```diff
@@ -117,6 +126,20 @@ func (g Generator) Generate(ctx *genall.GenerationContext) error {
 			}
 		}
 
+		// *If* we're only generating v1beta1 CRDs, default to `preserveUnknownFields: (unset)`
+		// for compatibility purposes.  In any other case, default to false, since that's
+		// the sensible default and is required for v1.
+		v1beta1Only := len(crdVersions) == 1 && crdVersions[0] == "v1beta1"
+		switch {
+		case (g.PreserveUnknownFields == nil || *g.PreserveUnknownFields) && v1beta1Only:
+			crd := versionedCRDs[0].(*apiextlegacy.CustomResourceDefinition)
+			crd.Spec.PreserveUnknownFields = nil
+		case g.PreserveUnknownFields == nil, g.PreserveUnknownFields != nil && !*g.PreserveUnknownFields:
+			// it'll be false here (coming from v1) -- leave it as such
+		default:
+			return fmt.Errorf("you may only set PreserveUnknownFields to true with v1beta1 CRDs")
```

Comment:
It seems current behavior of this piece of code can be captured in the following table:

| initial value  | v1beta1  | v1  |   
|---|---|---|
| true  | nil  | not allowed  |   
| false  | **not allowed**  | false  |   
| nil  | nil  | nil  |   

I don't understand why if the initial value is false, we don't allowed it in v1beta1.
Per https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#pruning-versus-preserving-unknown-fields, it seems we can set it to false in v1beta1.

Reply:
your chart is off (which maybe means this code needs refactoring), but the second case in the switch statement handles this.

Reply:
basically, this code is roughly:

```rust
match g.PreserveUnknownFields {
  Some(true) if v1beta1Only => { crd.Spec.PreserveUnknownFields = nil }
  None if v1beta1Only => { crd.Spec.PreserveUnknownFields = nil }
  Some(false) => { /* nothing, false by default }
  None => { /* Nothing, false by default */ }
}
```

I'll see if I can make it more obvious

Reply:
Sorry for misreading it :P